laalaa/apps/advisers/tasks.py

```diff
@@ -95,33 +124,96 @@ class ImportProcess(Thread):
         'OUTREACH SERVICE',
     )
 
-    def __init__(self, xlsx_file, record, **options):
-        super(ImportProcess, self).__init__()
-        self.options = options
-        self.progress = {'task': 'initialising'}
-        self.record = record
-        self._interrupt = Event()
+    def __init__(self):
+        self.total = None
+        self.record = None
         self.sheets = {}
         self.temp_dir = tempfile.mkdtemp()
+        self.progress = {'task': 'initialising'}
 
-        if options.pop('single_transaction', True):
-            self.run = transaction.atomic(self.run)
+    def run(self, xlsx_file, record=None, *args, **kwargs):
+        cache.clear()
+        self.record = record
 
-        if options.pop('prime_geocoder', True):
-            prime_geocoder_cache()
+        self.update_state(
+            state='INITIALIZING',
+            meta=self.progress)
 
-        if options.pop('clear_db', True):
-            clear_db()
+        prime_geocoder_cache()
 
         csv_metadata = self.convert_excel_to_csv(xlsx_file)
         for csv_filename, headers, types in csv_metadata:
             self.load_csv_into_db(csv_filename, headers, types)
 
+        clear_db()
+
         self.translate_data()
 
         for meta in csv_metadata:
             self.drop_csv_table(meta[0])
 
+        cursor = connection.cursor()
+        cursor.execute("""
+            SELECT DISTINCT postcode FROM advisers_location""")
+        postcodes = cursor.fetchall()
+
+        self.total = len(postcodes)
+
+        def chunks(n=1000):
+            for i in xrange(0, len(postcodes), n):
+                yield postcodes[i:i + n]
+
+        self.update_count()
+
+        tasks = []
+        for chunk in chunks():
+            t = GeocoderTask().subtask(args=(chunk,))
+            tasks.append(t)
+        ts = TaskSet(tasks=tasks)
+        res = ts.apply_async()
+
+        task_counts = {}
+        task_errors = {}
+
+        def update_task_process(task_id, result):
+            task_counts[task_id] = result.get('count')
+            task_errors[task_id] = result.get('errors')
+
+        while res.completed_count() < len(tasks):
+            [update_task_process(r.task_id, r.result) for r in res
+             if r.result]
```

Comment:
Instead of using a list comprehension whose value is discarded, this would be clearer as a simple for loop:
```python
for r in res:
    if r.result:
        update_task_process(r.task_id, r.result)
```